components/server/src/api/auth-provider-service-api.ts

```diff
@@ -0,0 +1,201 @@
+            pagination: {
+                total: redacted.length,
+            },
+        });
+        return result;
+    }
+
+    /**
+     * Listing descriptions of auth providers doesn't require authentication.
+     */
+    @Unauthenticated()
+    async listAuthProviderDescriptions(
+        request: ListAuthProviderDescriptionsRequest,
+        context: HandlerContext,
+    ): Promise<ListAuthProviderDescriptionsResponse> {
+        const user = context.user;
+        const aps = user
+            ? await this.authProviderService.getAuthProviderDescriptions(user)
+            : await this.authProviderService.getAuthProviderDescriptionsUnauthenticated();
+
+        const selectedProviders = selectPage(aps, request.pagination);
+        return new ListAuthProviderDescriptionsResponse({
+            descriptions: selectedProviders.map((ap: AuthProviderInfo) =>
+                this.apiConverter.toAuthProviderDescription(ap),
+            ),
+        });
+    }
+
+    async updateAuthProvider(
+        request: UpdateAuthProviderRequest,
+        context: HandlerContext,
+    ): Promise<UpdateAuthProviderResponse> {
+        if (!request.authProviderId) {
+            throw new ConnectError("authProviderId is required", Code.InvalidArgument);
+        }
+        const clientId = request.clientId;
+        const clientSecret = request.clientSecret;
+        if (!clientId || typeof clientSecret === "undefined") {
+            throw new ConnectError("clientId or clientSecret are required", Code.InvalidArgument);
+        }
+
+        const authProvider = await this.authProviderService.getAuthProvider(context.user.id, request.authProviderId);
+        if (!authProvider) {
+            throw new ConnectError("Provider not found.", Code.NotFound);
+        }
+
+        let entry: AuthProviderEntry;
+        if (authProvider.organizationId) {
+            entry = await this.authProviderService.updateOrgAuthProvider(context.user.id, {
+                id: request.authProviderId,
+                organizationId: authProvider.organizationId,
+                clientId: clientId,
+                clientSecret: clientSecret,
+            });
+        } else {
+            entry = await this.authProviderService.updateAuthProviderOfUser(context.user.id, {
+                id: request.authProviderId,
+                ownerId: context.user.id,
+                clientId: clientId,
+                clientSecret: clientSecret,
+            });
+        }
+
+        return new UpdateAuthProviderResponse({
+            authProvider: this.apiConverter.toAuthProvider(AuthProviderEntry.redact(entry)),
+        });
+    }
+
+    async deleteAuthProvider(
+        request: DeleteAuthProviderRequest,
+        context: HandlerContext,
+    ): Promise<DeleteAuthProviderResponse> {
+        if (!request.authProviderId) {
+            throw new ConnectError("authProviderId is required", Code.InvalidArgument);
+        }
+
```

Comment:
Should not we have a check that either organizationId or userId is valid?

Reply:
> message DeleteAuthProviderRequest { string auth_provider_id = 1; }

Given that's the params, what exactly you want to check there? FGA should check for the subject's permissions here.

Reply:
sorry!